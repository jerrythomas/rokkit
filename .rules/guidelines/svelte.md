# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
