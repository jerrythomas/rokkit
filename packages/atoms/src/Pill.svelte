<script>
	import { FieldMapper, noop } from '@rokkit/core'
	import Item from './Item.svelte'
	import Icon from './Icon.svelte'
	import { keyboard } from '@rokkit/actions'

	/**
	 * @type {Object} Props
	 * @property {string|Object} [value]
	 * @property {FieldMapper}   [mapping]
	 * @property {boolean} [removable]
	 * @property {boolean} [disabled]
	 */

	/** @type {Props} */
	let {
		value,
		mapping = new FieldMapper(),
		removable = false,
		disabled = false,
		onremove = noop
	} = $props()

	const keyMappings = {
		remove: ['Delete', 'Backspace']
	}

	function handle(event) {
		if (!disabled) onremove(value)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<rk-pill use:keyboard={keyMappings} onremove={handle} tabindex="0">
	<Item {value} {mapping}></Item>
	{#if removable}
		<button {disabled} onclick={handle}>
			<Icon name="remove" />
		</button>
	{/if}
</rk-pill>
