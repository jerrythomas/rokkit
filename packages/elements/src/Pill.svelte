<script>
	import { FieldMapper, noop } from '@rokkit/core'
	import Item from './Item.svelte'
	import Icon from './Icon.svelte'
	import { keyboard } from '@rokkit/actions'

	/**
	 * @type {Object} Props
	 * @property {string|Object} [value]
	 * @property {FieldMapper}   [mapping]
	 * @property {boolean} [removable]
	 * @property {boolean} [disabled]
	 */

	/** @type {Props} */
	let {
		value,
		class: classes = '',
		mapping = new FieldMapper(),
		removable = false,
		disabled = false,
		onremove = noop
	} = $props()

	const keyMappings = {
		remove: ['Delete', 'Backspace']
	}

	function handle(event) {
		if (!disabled) onremove(value)
	}
</script>

<!-- svelte-ignore a11y_no_noninteractive_tabindex -->
<rk-pill use:keyboard={keyMappings} onremove={handle} tabindex="0" class={classes}>
	<Item {value} {mapping}></Item>
	{#if removable}
		<Icon name="action-close" role="button" aria-label="Remove" {disabled} onclick={handle} small
		></Icon>
	{/if}
</rk-pill>
