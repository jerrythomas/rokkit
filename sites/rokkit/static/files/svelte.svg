<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
