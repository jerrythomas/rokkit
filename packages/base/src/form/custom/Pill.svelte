<script>
	import { createEventDispatcher } from 'svelte'
	const dispatch = createEventDispatcher()

	export let item
	export let fields

	function handleClick() {
		dispatch('remove', item)
	}
</script>

<pill class="flex flex-row items-center">
	{item[fields.text]}
	<icon class="remove" on:click={handleClick} />
</pill>
