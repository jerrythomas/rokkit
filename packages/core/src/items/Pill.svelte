<script>
	import { createEventDispatcher } from 'svelte'
	import { defaultFields, defaultStateIcons } from '../constants'
	import Text from './Text.svelte'
	const dispatch = createEventDispatcher()

	export let value
	export let fields = defaultFields
	export let using = { default: Text }
	export let removable = false

	function handleClick() {
		dispatch('remove', value)
	}
	$: component =
		typeof value == 'object'
			? using[value[fields.component] ?? 'default']
			: using.default
</script>

<pill class="flex flex-row items-center">
	<item class="flex flex-row items-center">
		<svelte:component this={component} content={value} {fields} />
	</item>
	<!-- svelte-ignore a11y-click-events-have-key-events -->
	{#if removable}
		<icon
			class="{defaultStateIcons.action.remove} cursor-pointer"
			on:click={handleClick}
		/>
	{/if}
</pill>
