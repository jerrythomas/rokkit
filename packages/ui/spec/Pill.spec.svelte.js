import { describe, it, expect, vi } from 'vitest'
import { render, fireEvent } from '@testing-library/svelte'
import { flushSync, tick } from 'svelte'
import Pill from '../src/Pill.svelte'

describe('Pill', () => {
	it('should render', () => {
		const props = $state({ value: 'Hello' })
		const { container } = render(Pill, { props })
		expect(container).toMatchSnapshot()

		props.value = 'World'
		flushSync()
		expect(container).toMatchSnapshot()
	})

	it('should render with icon', () => {
		const props = $state({ value: { text: 'Hello', icon: 'world' } })
		const { container } = render(Pill, { props })
		expect(container).toMatchSnapshot()

		props.value = { text: 'World', icon: 'hello' }
		flushSync()
		expect(container).toMatchSnapshot()
	})

	it('should render with image', () => {
		const props = $state({ value: { text: 'Hello', image: 'world.png' } })
		const { container } = render(Pill, { props })
		expect(container).toMatchSnapshot()

		props.value = { text: 'World', image: 'hello.png' }
		flushSync()
		expect(container).toMatchSnapshot()
	})

	it('should fire remove event on click', async () => {
		const props = $state({ value: 'Hello', removable: true, disabled: true, onremove: vi.fn() })
		const { container } = render(Pill, { props })

		expect(container).toMatchSnapshot()
		const closeButton = container.querySelector('rk-icon')

		fireEvent.click(closeButton)
		await tick()
		expect(props.onremove).not.toHaveBeenCalled()

		props.disabled = false
		flushSync()
		expect(container).toMatchSnapshot()
		fireEvent.click(closeButton)
		await tick()
		expect(props.onremove).toHaveBeenCalledWith('Hello')
	})

	it.each(['Delete', 'Backspace'])('should fire remove event on %s', async (key) => {
		const props = $state({ value: 'Hello', removable: true, disabled: true, onremove: vi.fn() })
		const { container } = render(Pill, { props })
		expect(container).toMatchSnapshot()
		const pill = container.querySelector('rk-pill')

		fireEvent.keyUp(pill, { key })
		await tick()
		expect(props.onremove).not.toHaveBeenCalled()

		props.disabled = false
		flushSync()
		expect(container).toMatchSnapshot()
		fireEvent.keyUp(pill, { key })
		expect(props.onremove).toHaveBeenCalledWith('Hello')
	})
})
