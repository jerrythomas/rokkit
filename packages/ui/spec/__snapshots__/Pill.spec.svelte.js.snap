// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pill > should fire remove event on Backspace 1`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <rk-icon
      aria-label="action-close"
      data-disabled="true"
      data-size="base"
      data-tag-button="true"
      data-tag-icon="true"
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="action-close"
      />
    </rk-icon>
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should fire remove event on Backspace 2`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <rk-icon
      aria-label="action-close"
      data-disabled="false"
      data-size="base"
      data-tag-button="true"
      data-tag-icon="true"
      role="button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="action-close"
      />
    </rk-icon>
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should fire remove event on Delete 1`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <rk-icon
      aria-label="action-close"
      data-disabled="true"
      data-size="base"
      data-tag-button="true"
      data-tag-icon="true"
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="action-close"
      />
    </rk-icon>
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should fire remove event on Delete 2`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <rk-icon
      aria-label="action-close"
      data-disabled="false"
      data-size="base"
      data-tag-button="true"
      data-tag-icon="true"
      role="button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="action-close"
      />
    </rk-icon>
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should fire remove event on click 1`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <rk-icon
      aria-label="action-close"
      data-disabled="true"
      data-size="base"
      data-tag-button="true"
      data-tag-icon="true"
      role="button"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="action-close"
      />
    </rk-icon>
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should fire remove event on click 2`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <rk-icon
      aria-label="action-close"
      data-disabled="false"
      data-size="base"
      data-tag-button="true"
      data-tag-icon="true"
      role="button"
      tabindex="0"
    >
      <i
        aria-hidden="true"
        class="action-close"
      />
    </rk-icon>
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should render 1`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should render 2`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <!---->
     
    <p>
      World
    </p>
    <!---->
    <!---->
     
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should render with icon 1`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <rk-icon
      aria-label="Hello"
      data-disabled="false"
      data-size="base"
      data-tag-icon="true"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="world"
      />
    </rk-icon>
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should render with icon 2`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <rk-icon
      aria-label="World"
      data-disabled="false"
      data-size="base"
      data-tag-icon="true"
      role="img"
      tabindex="-1"
    >
      <i
        aria-hidden="true"
        class="hello"
      />
    </rk-icon>
    <!---->
     
    <p>
      World
    </p>
    <!---->
    <!---->
     
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should render with image 1`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <img
      alt="Hello"
      src="world.png"
    />
    <!---->
     
    <p>
      Hello
    </p>
    <!---->
    <!---->
     
    <!---->
  </rk-pill>
  
</div>
`;

exports[`Pill > should render with image 2`] = `
<div>
  <rk-pill
    tabindex="0"
  >
    <!---->
    <img
      alt="World"
      src="hello.png"
    />
    <!---->
     
    <p>
      World
    </p>
    <!---->
    <!---->
     
    <!---->
  </rk-pill>
  
</div>
`;
